File: week-6-js-exercises/products-collection-oop/products-collection.js
function ProductList(products) {
    this.products = products;
}

ProductList.prototype.count = function() {
    return this.products.length;
};

ProductList.prototype.getNames = function () {
    return this.products.map(function(product) {
        return product.name;
    });
};

ProductList.prototype.filterByBrand = function (brand) {
    return this.products.filter(function(product) {
        return product.brand === brand;
    });
};

ProductList.prototype.countByBrand = function (brand) {
    return this.filterByBrand(brand).length;
};

ProductList.prototype.groupByBrand = function (brand) {
    // var results = {};
    // this.products.forEach(function(product) {
    //     if (!results[product.brand]) {
    //         results[product.brand] = [];
    //     }
    //     results[product.brand].push(product);
    // });

    return this.products.reduce(function(accumulator, product) {
        if (!accumulator[product.brand]) {
            accumulator[product.brand] = [];
        }
        accumulator[product.brand].push(product);
        return accumulator;
    }, {});
};


ProductList.prototype.cheapest = function (price) {
    return this.Math.min.apply( );
};

var list = new ProductList([
    { name: "Samsung Galaxy S6 Edge", price: 829.90, brand: "Samsung" },
    { name: "LG G Flex", price: 689.90, brand: "LG" },
    { name: "Apple iPhone 6 Plus", price: 779.90, brand: "Apple" },
    { name: "Samsung Galaxy Alpha", price: 589.90, brand: "Samsung" },
    { name: "Nokia Lumia 930", price: 489.90, brand: "Nokia" },
    { name: "Sony Xperia Z3 Compact", price: 389.90, brand: "Sony" },
    { name: "Apple iPhone 6", price: 779.90, brand: "Apple" },
    { name: "Huawei Ascend G7", price: 259.90, brand: "Huawei" },
    { name: "Sony Xperia T2 Ultra", price: 359.90, brand: "Sony" },
    { name: "LG G4", price: 589.90, brand: "LG" },
    { name: "Samsung Galaxy Note 4", price: 689.90, brand: "Samsung" },
    { name: "HTC One M9", price: 674.90, brand: "HTC" },
    { name: "Apple iPhone 5S", price: 589.90, brand: "Apple" }
]);

// console.log(list.count());
// console.log(list.getNames());
// console.log(list.filterByBrand("LG"));
// console.log(list.countByBrand("LG"));
console.log(list.groupByBrand());

File: week-7-ruby-exercises/products-collection-oop-tdd/products_collection.rb
class Product
  attr_reader :name, :price, :brand

  def initialize(options)
    @name = options[:name]
    @price = options[:price]
    @brand = options[:brand]
  end

  def to_s
    "#{@name}, €#{@price}"
  end
end

class ProductList
  def initialize(products)
    @products = products.map do |product|
      Product.new(product)
    end
  end

  def count
    @products.size
  end

  def get_names
    @products.map do |product|
      product.name
    end
  end

  def filter_by_brand(brand)
    @products.select do |product|
      product.brand == brand
    end
  end

  def count_by_brand(brand)
    filter_by_brand(brand).size
  end

  def group_by_brand
    @products.group_by do |product|
      product.brand
    end
  end

  def cheapest
    @products.min do |a, b|
      a.price <=> b.price
    end
  end

  def most_expensive
    @products.max do |a, b|
      a.price <=> b.price
    end
  end

  def price_between(min, max)
    @products.reject do |product|
      product.price < min || product.price > max
    end
  end

  def avg_price
    total = @products.reduce(0) do |memo, product|
      memo += product.price
    end
    (total / @products.size).round(2)
  end
end
